### PR TITLE
Cache best trial_id

### PR DESCRIPTION
**Pull request by oda**
Hi, I made a cache for the in-memory storage.

### CODE REVIEW

**Comment by sile** [comment]
@g-votte @toshihikoyanase Please review this PR before merging.

File: optuna/storages/in_memory.py [hunk 1]
```diff
@@ -236,6 +247,31 @@ def set_trial_value(self, trial_id, value):
 
             self.trials[trial_id] = self.trials[trial_id]._replace(value=value)
 
+    def update_cache(self, trial_id):
```
Comment (toshihikoyanase) [inline optuna/storages/in_memory.py]: Please make this method private by adding `_` to the head of the name. I think this method is used only in this file.

File: optuna/storages/in_memory.py [hunk 2]
```diff
@@ -236,6 +247,31 @@ def set_trial_value(self, trial_id, value):
...
+            return
+        if new_value is None:
+            return
+        if (self.get_study_direction(IN_MEMORY_STORAGE_STUDY_ID) ==
+                structs.StudyDirection.MAXIMIZE):
+            if best_value < new_value:
+                self.best_trial_id = trial_id
+            return
+        if best_value > new_value:
+            self.best_trial_id = trial_id
```
Comment (toshihikoyanase) [inline optuna/storages/in_memory.py]: Please add test cases to confirm that `self.best_trial_id` is updated correctly whether the study direction is `MINIMIZE` or `MAXIMIZE`.

File: optuna/storages/in_memory.py (hunk 1 above)
Comment (toshihikoyanase) [reply optuna/storages/in_memory.py]: The method will be renamed in #611.

File: optuna/storages/in_memory.py (hunk 2 above)
Comment (toshihikoyanase) [reply optuna/storages/in_memory.py]: The test cases will be added in #611.

**Pull request merged**